pkg/volume/csi/csi_mounter.go

```diff
@@ -197,7 +197,23 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 	}
 
 	// apply volume ownership
-	if !c.readOnly && fsGroup != nil {
+	// The following logic is derived from https://github.com/kubernetes/kubernetes/issues/66323
+	// if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
+	// if fstype is provided and pv.AccessMode == ReadWriteOnly, then apply fsgroup
+
+	if fsGroup != nil {
+		if fsType == "" {
+			glog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, fsType not provided"))
+			return nil
+		}
+
+		accessModes := c.spec.PersistentVolume.Spec.AccessModes
+		if !hasReadWriteOnce(accessModes) {
+			glog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, only applies to ReadWriteOnce access mode"))
+			return nil
+		}
+
+		glog.V(4).Info(log("mounter.SetupAt applying fsGroup %d...", *fsGroup))
```

Comment:
Somewhat minor nit - typically when debugging an issue, I look for `mounter.SetUp successfully` (with correct log level of course) but it seems weird to me that, that message will only be logged when device is of type block storage with file system otherwise the message will not appear in logs at all.

As someone who has spent lot of time debugging/reading logs - it would be nice to maintain that symmetry. That message is logged below btw. If it helps the entire volume ownership stuff could be moved to a new function and called from here, that way even if that function returns without doing anything. Our `mounter.setup successfully` message will always be logged.

Reply:
Makes sense. I will rework to make sure proper messaging is logged.  Thanks!

Reply:
@gnufied i checked the code, the `mouter.Setup successfully` message is outside the `if-statement` block for the fsgroup logic.  Meaning, it will show up regardless of whether fsgroup was applied or not.  Let me know if I am missing something here.

Reply:
I meant we are returning early from the function without logging `mouter.Setup successfully` when there is no need to change owership right? So the statement outside if will never be executed.